curl-helper.c

```diff
@@ -4862,13 +4839,44 @@ value caml_curl_multi_poll(value v_timeout_ms, value v_extra_fds, value v_multi)
   CAMLreturn(Val_bool(numfds != 0));
 }
 
+static int curlm_closesocket_cb(void *data, curl_socket_t socket)
+{
+  caml_leave_blocking_section();
+
+  CAMLparam0();
+  CAMLlocal1(camlResult);
+
+  ml_multi_handle* multi = (ml_multi_handle*)data;
+  int result = 0;
+  camlResult = caml_callback_exn(Field(multi->values, curlmopt_closesocket_function), Val_socket(socket));
+  if (Is_exception_result(camlResult))
+  {
+    result = 1;
+  }
+  CAMLdrop;
+
+  caml_enter_blocking_section();
+  return result;
+}
+
 value caml_curl_multi_add_handle(value v_multi, value v_easy)
 {
   CAMLparam2(v_multi,v_easy);
   CURLMcode rc = CURLM_OK;
   CURLM* multi = CURLM_val(v_multi);
   Connection* conn = Connection_val(v_easy);
 
+  if (Field(Multi_val(v_multi)->values, curlmopt_closesocket_function) != Val_unit)
+  {
+    CURLcode result = CURLE_OK;
+    result = curl_easy_setopt(conn->handle, CURLOPT_CLOSESOCKETFUNCTION, curlm_closesocket_cb);
+    if (result != CURLE_OK)
+      raiseError(conn, result);
+    result = curl_easy_setopt(conn->handle, CURLOPT_CLOSESOCKETDATA, Multi_val(v_multi));
+    if (result != CURLE_OK)
+      raiseError(conn, result);
```

Comment:
I think that it would be better to invert the two calls to `curl_easy_setopt`; otherwise if for some reason the second call fails and the callback is called, the `data` argument will be invalid. Alternatively, the callback should be unset if the second call fails.

Reply:
Good idea, done.